## ひろしまQuest2020#stayhome【球種予測部門】

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 100)

In [2]:
ALL_MERGE = 'intermediate/all_merge_12.f'
SUBMIT = 'submit/ball_12.csv'

In [3]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 171)

### 不要な列を削除

In [4]:
all_pitch.drop(
    columns=[
        '年度', 
        '試合ID', 
        'ホームチームID', 'アウェイチームID', 
        '投手ID', '投手チームID', 
        '打者ID', '打者チームID', 
        'プレイ前走者状況', 
        '捕手ID', 
        'opening_date', 'game_date',
        'start_time', 'game_time', 'elapsed_time'
    ], inplace=True)

### Rename

In [5]:
all_pitch.rename(columns={
    'データ内連番': 'No',
    '投球位置区域': 'pitch_area',
    '試合内連番': 'No_in_game',
    '試合内投球数': 'pitch_cnt_in_game',
    'イニング': 'inning',
    'イニング内打席数': 'bat_cnt_in_inning',
    '打席内投球数': 'pitch_cnt_in_bat',
    '投手登板順': 'pitch_order',
    '投手試合内対戦打者数': 'player_cnt_in_game',
    '投手試合内投球数': 'pitcher_cnt_in_game',
    '投手イニング内投球数': 'pitcher_cnt_in_inning',
    '打者打順': 'bat_order',
    '打者試合内打席数': 'bat_cnt_in_game',
    'プレイ前ホームチーム得点数': 'home_point',
    'プレイ前アウェイチーム得点数': 'away_point',
    'プレイ前アウト数': 'out_cnt',
    'プレイ前ボール数': 'ball_cnt',
    'プレイ前ストライク数': 'strike_cnt',
}, inplace=True)

In [6]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 156)


No  ball  pitch_area  No_in_game  pitch_cnt_in_game  inning  \
0   0   0.0         0.0           1                  1       1   
1   1   0.0         8.0           2                  2       1   
2   2   0.0         5.0           3                  3       1   
3   3   0.0        12.0           4                  4       1   
4   4   0.0         8.0           5                  5       1   

   bat_cnt_in_inning  pitch_cnt_in_bat  pitch_order  player_cnt_in_game  \
0                  1                 1            1                   1   
1                  1                 2            1                   1   
2                  1                 3            1                   1   
3                  2                 1            1                   2   
4                  2                 2            1                   2   

   pitcher_cnt_in_game  pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  \
0                    1                      1          1                1   
1                    2                      2          1                1   
2                    3                      3          1                1   
3                    4                      4          2                1   
4                    5                      5          2                1   

   home_point  away_point  out_cnt  ball_cnt  strike_cnt  bc_straight  \
0           0           0        0         0           0     0.477023   
1           0           0        0         0           1     0.459832   
2           0           0        0         0           2     0.469730   
3           0           0        1         0           0     0.477023   
4           0           0        1         0           1     0.459832   

   bc_curve  bc_slider  bc_shoot   bc_fork  bc_changeup  bc_sinker  \
0  0.097592   0.200870  0.074454  0.046799     0.041882   0.008619   
1  0.090366   0.161893  0.068313  0.099045     0.061079   0.012505   
2  0.066062   0.159395  0.031361  0.162940     0.062722   0.014112   
3  0.097592   0.200870  0.074454  0.046799     0.041882   0.008619   
4  0.090366   0.161893  0.068313  0.099045     0.061079   0.012505   

   bc_cutball  bc_area0  bc_area1  bc_area2  bc_area3  bc_area4  bc_area5  \
0    0.052760  0.048489  0.071105  0.059259  0.047675  0.063792  0.049150   
1    0.046967  0.035008  0.056450  0.047706  0.035812  0.044074  0.034687   
2    0.033679  0.021884  0.035656  0.030884  0.020316  0.021748  0.021816   
3    0.052760  0.048489  0.071105  0.059259  0.047675  0.063792  0.049150   
4    0.046967  0.035008  0.056450  0.047706  0.035812  0.044074  0.034687   

   bc_area6  bc_area7  bc_area8  bc_area9  bc_area10  bc_area11  bc_area12  \
0  0.046338  0.070075  0.059013  0.117934   0.090325   0.114800   0.162045   
1  0.037484  0.058379  0.050407  0.112451   0.103867   0.161700   0.221976   
2  0.028429  0.036678  0.029520  0.133761   0.131170   0.201595   0.286542   
3  0.046338  0.070075  0.059013  0.117934   0.090325   0.114800   0.162045   
4  0.037484  0.058379  0.050407  0.112451   0.103867   0.161700   0.221976   

   first  second  third  base_cnt  pitch_LR  bat_LR  top_bot  role  pos_pit  \
0      0       0      0         0       0.0     1.0        1     1        0   
1      0       0      0         0       0.0     1.0        1     1        0   
2      0       0      0         0       0.0     1.0        1     1        0   
3      0       0      0         0       0.0     0.0        1     1        0   
4      0       0      0         0       0.0     0.0        1     1        0   

   date_from_opening  ...  firm_bat  height_bat  weight_bat  draft_order_bat  \
0                  0  ...         0         176          74              5.0   
1                  0  ...         0         176          74              5.0   
2                  0  ...         0         176          74              5.0   
3                  0  ...         0         180          74              1.0   
4                  0  ...

### train
- 行数: 257117

In [7]:
train = all_pitch.dropna(subset=['pitch_area'])
train.shape

(257117, 156)

### test
- 行数: 521650

In [8]:
test = all_pitch[all_pitch['pitch_area'].isnull()]
test.shape

(521650, 156)

In [9]:
del all_pitch
gc.collect()

20

In [10]:
train_d = train.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

In [11]:
test_d = test.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

## Dataset作成

In [12]:
lgb_train = lgb.Dataset(train_d, train['ball'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [13]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 8,
        'seed' : 0,
        'learning_rate' : 0.1,
        'lambda_l1': 6.9923570049658075, 
        'lambda_l2': 0.002378623984798833, 
        'num_leaves': 18, 
        'feature_fraction': 0.45199999999999996, 
        'bagging_fraction': 0.9799724836460725, 
        'bagging_freq': 4, 
        'min_child_samples': 20
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
best_cv_score = cv_results['multi_logloss-mean'][-1]
print('Best CV score:', best_cv_score)

[100]	cv_agg's multi_logloss: 1.15178 + 0.00193454
[200]	cv_agg's multi_logloss: 1.12973 + 0.00215014
[300]	cv_agg's multi_logloss: 1.12218 + 0.0022383
[400]	cv_agg's multi_logloss: 1.11768 + 0.00223602
[500]	cv_agg's multi_logloss: 1.11446 + 0.00239005
[600]	cv_agg's multi_logloss: 1.11205 + 0.0025788
[700]	cv_agg's multi_logloss: 1.11036 + 0.00271337
[800]	cv_agg's multi_logloss: 1.10909 + 0.00280035
[900]	cv_agg's multi_logloss: 1.10808 + 0.00297999
[1000]	cv_agg's multi_logloss: 1.1076 + 0.00302675
[1100]	cv_agg's multi_logloss: 1.1072 + 0.00297822
[1200]	cv_agg's multi_logloss: 1.10706 + 0.00294847
[1300]	cv_agg's multi_logloss: 1.107 + 0.00293785
[1400]	cv_agg's multi_logloss: 1.10706 + 0.00302059
Best num_boost_round: 1353
Best CV score: 1.106942086205103
CPU times: user 1h 45min 8s, sys: 58.9 s, total: 1h 46min 7s
Wall time: 27min 41s


In [14]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [15]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 32min 50s, sys: 33.8 s, total: 33min 23s
Wall time: 9min 2s


### Feature Importance

In [16]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
144    salary_year_dif_p-c      2308
149            age_dif_b-c      2312
117      salary_x_year_bat      2387
137            age_dif_p-b      2449
105             weight_bat      2534
81               curve_pit      2552
104             height_bat      2580
84                fork_pit      2599
6       player_cnt_in_game      2651
146            bmi_dif_p-c      2754
116        salary_year_bat      2781
135         salary_dif_p-b      2808
110       bat_game_cnt_bat      2811
139  salary_x_year_dif_p-b      2837
147         salary_dif_b-c      2865
52              point_diff      2918
151  salary_x_year_dif_b-c      2980
118                bmi_bat      3061
82              slider_pit      3245
138    salary_year_dif_p-b      3405
109         batter_cnt_bat      3639
150    salary_year_dif_b-c      3739
0               No_in_game      4047
1        pitch_cnt_in_game      4239
152            bmi_dif_b-c      4338
140            bmi_dif_p-b      4395
8    pitcher_cnt_in_inning      4524
47             elapsed_min      4770
7      pitcher_cnt_in_game      5295
46       date_from_opening      7187

In [17]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 17min 21s, sys: 4.16 s, total: 17min 25s
Wall time: 4min 35s


## 結果出力

In [18]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 9)


index         0         1         2         3         4         5  \
0      0  0.404815  0.026219  0.134029  0.319365  0.001857  0.000016   
1      1  0.396373  0.069012  0.109905  0.362815  0.003301  0.000020   
2      2  0.360386  0.068595  0.136540  0.185437  0.151565  0.000090   
3      3  0.355107  0.098545  0.188993  0.271665  0.015234  0.000044   
4      4  0.203164  0.048151  0.308949  0.315412  0.009748  0.000042   

          6         7  
0  0.000014  0.113686  
1  0.000010  0.058564  
2  0.000030  0.097358  
3  0.000018  0.070394  
4  0.000016  0.114517

### 2017の平均
- ストレート           0.468254
- カーブ             0.074725
- スライダー           0.185806
- シュート            0.070633
- フォーク            0.083013
- チェンジアップ         0.054217
- シンカー            0.011361
- カットボール          0.051992

In [19]:
submit.mean()

index    260824.500000
0             0.527723
1             0.076577
2             0.178028
3             0.052802
4             0.070016
5             0.042633
6             0.007617
7             0.044603
dtype: float64

In [20]:
submit.to_csv(SUBMIT, header=False, index=False)

### SIGNATE

In [21]:
'signate submit --competition-id=275 ~/work/baseball/{} --note feat={}_cv={}'.format(SUBMIT, column_cnt, best_cv_score)

'signate submit --competition-id=275 ~/work/baseball/submit/ball_12.csv --note feat=156_cv=1.106942086205103'

### GCS

In [22]:
'gsutil cp ~/work/baseball/{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ~/work/baseball/submit/ball_12.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 12 (156) ... コースの集計を追加
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1353
- Best CV score: 1.106942086205103
- Wall time: 27min 41s
- 評価結果  : 1.4016077

#### 11 (143) ... ボールカウントの集計を追加
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1209
- Best CV score: 1.1075153721728936
- Wall time: 23min 19s
- 評価結果  : 未提出

#### 10_tune (135) ... 再チューニング
- 'num_leaves' : 15,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1260
- Best CV score: 1.1118897437428692
- Wall time: 8min 40s
- 評価結果  : 1.4060397

#### 10s (135) ... trainを75%サンプリング
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1412
- Best CV score: 1.1190287120710383
- Wall time: 12min 5s
- 評価結果  : 1.4310869

#### 10 (135) ... 5相当に戻した
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1112
- Best CV score: 1.111265000190272
- Wall time: 9min 30s
- 評価結果  : 1.4008712

#### 9 (136) ... 5相当に戻した=>NG
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1106
- Best CV score: 1.1113177029272225
- Wall time: 25min 44s (local)
- 評価結果  : 1.5133521

#### 8 (157)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1100
- Best CV score: 1.1093728644236824
- Wall time: 10min 4s
- 評価結果  : 1.4925832

#### 7 (235)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1204
- Best CV score: 1.108590082437519
- Wall time: 12min 56s
- 評価結果  : 1.4946937

#### 6 (207)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1135
- Best CV score: 1.1086837935965734
- Wall time: 11min 54s
- 評価結果  : 1.3999194

#### 5-tune (参考)
- 'num_leaves' : 18,
- 'learning_rate' : 0.05,
- Best num_boost_round: 2395
- Best CV score: 1.1103399384119979
- Wall time: 18min 29s (n1-standard-8)

#### 5-tune (135)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- CV score: 1.1112402566616455
- 評価結果  : 1.3991780
- CV Wall time: 27min 48s (local) -> 18min 24s (n1-standard-4) -> 9min 3s (n1-standard-8)

#### 5 (135)
- 'num_leaves' : 31,
- 'learning_rate' : 0.1,
- CV score: 1.1161825304353852
- 評価結果  : 1.4252286

#### 4 (128)
- CV score: 1.1158111693845418
- 評価結果  : 1.4299265

#### 3 (97)
- CV score: 1.1172290313076036
- 評価結果  : 1.4335806